### DISCUSSION TITLE
NoCoLogging with custom items

### DISCUSSION THREAD

**Discussion by mainrs** [klogging/klogging#221] [Q&A]
Something like `withLogContext` but for non-coroutine contexts. Is that possible?

**Comment by mjstrasser** [comment]
Hi @mainrs, thanks for your question.

Do you mean retrieving coroutine context items from a `NoCoLogger` instance? I cannot see an easy way to do that. All the calls in a stack must be to suspend functions in order to pass coroutine context. If you “break the thread” by calling a non-suspend function, that function and any others it calls lose access to the context.

Java logging frameworks include thread-based context storage, e.g. [Log4j Thread Context](https://logging.apache.org/log4j/2.x/manual/thread-context.html) and [Logback Mapped Diagnostic Context (MDC)](https://logback.qos.ch/manual/mdc.html). These store context items in Java `ThreadLocal` maps that can be retrieved during log event processing. If code execution is moved between threads and Thread Context or MDC entries need to be maintained, it is the responsibility of the programmers to ensure the entries are moved as well. Spring, Quarkus and probably other frameworks take care of this work.

If you use [SLF4J](https://slf4j.org/) and the [Klogging SL4J Provider](https://klogging.io/docs/java/slf4j), all SLF4J MDC entries are included in Klogging log events when they are emitted. There is a [small example](https://klogging.io/docs/java/slf4j#mdc) in the Klogging docs for SLF4J. This applies to all Klogging loggers, whether created by the SLF4J Provider or directly using Klogging.

Does this answer your question? Let me know if you need more information.

**Comment by mainrs** [comment]
I was hoping for something like `emitEvent(level, throwable?, event, items)` but with ergonomic functions like `info`, `trace` etc. The base function already exists to accomplish this. The utility functions around it are not available.

**Reply by mjstrasser** [reply]
OK, I didn’t understand that from your original question.

Can you give me some examples of what you would like to see and how you would use it? You are welcome to create a PR with suggested enhancements.

**Reply by mainrs** [reply]
Probably something like this:

```kotlin
// News reader with offline support scenario.

val locallyStoredNewsArticles: List<ArticleId> = databaseRepository.fetchLocallyStoredNewsArticleIds()
val newNewsArticles: List<Article> = remoteRepository.fetchNewsArticles()

val (existingArticles, newArticles) = newNewsArticles.partition { it.id in locallyStoredNewsArticles }

// Logging database transaction information.
val items = mapOf(
    "localNewsArticleCount" to locallyStoredNewsArticles.size,
    "remoteNewsArticleCount" to newNewsArtclees.size,
    "updatingArticlesCount" to existingArticles.size,
    "insertingArticlesCount" to newArticles,size,
)
logger.emitEvent(Level.INFO, null, "updating local database", items)

// Do action for inserting new articles.

// Do action for updating existing local articles.
```

I could use message templates. The problem with those is that the log message would require to contain all the variables I want to have in the log event. I personally am not a fan of the message template format, since the variable names are loosely typed. Compared to native Kotlin string interpolation for example. That's why I'd prefer a function where I can simply pass some items to.

**Reply by mjstrasser** [reply]
Thanks, that makes sense. 

I will implement this ability for both `Klogger` and `NoCoLogger`. It will probably take a week or so because I want to ensure the code does not get messy as I add new functionality.

**Reply by mjstrasser** [reply]
I have implemented what I call [immediate context items](https://klogging.io/docs/logging/sending-logs#immediate-context-items) for both `Klogger` and `NoCoLogger` in [release 0.5.9](https://github.com/klogging/klogging/releases/tag/0.5.9).

Let me know if this meets your needs.
